src/box/alter.cc

```diff
@@ -3913,15 +4024,69 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
 			/* Version can be major.minor with no patch. */
 			if (tuple_field_u32(new_tuple, 3, &patch) != 0)
 				patch = 0;
-			dd_version_id = version_id(major, minor, patch);
+			version = version_id(major, minor, patch);
 		} else {
 			assert(old_tuple != NULL);
 			/*
 			 * _schema:delete({'version'}) for
 			 * example, for box.internal.bootstrap().
 			 */
-			dd_version_id = tarantool_version_id();
+			version = tarantool_version_id();
+		}
+		struct trigger *on_commit = txn_alter_trigger_new(
+			on_commit_dd_version, (void *)(uintptr_t)version);
+		if (on_commit == NULL)
+			return -1;
+		txn_stmt_on_commit(stmt, on_commit);
+	} else if (strcmp(key, "cluster_name") == 0) {
+		char name[NODE_NAME_SIZE_MAX];
+		const char *field_name = "_schema['cluster_name'].value";
+		if (tuple_field_node_name(name, new_tuple,
+					  BOX_SCHEMA_FIELD_VALUE,
+					  field_name) != 0)
+			return -1;
+		if (box_is_configured() && strcmp(name, CLUSTER_NAME) != 0) {
```

Comment:
I suggest you allow setting CLUSTER_NAME and REPLICASET_NAME from empty to non-empty without force-recovery.

This will simplify the most common use case (adding names to an existing replicaset and cluster): basically, the user will issue `box.cfg{cluster_name = 'smth'}` on master of an existing replicaset, and the name will change and the change will propagate to everyone without problems and without the need to set force_recovery on every instance.

Any other rename (from one non-empty name to other) should still require force_recovery, of course.

Reply:
You have to realize that we still can't allow that at the first `box.cfg{}`. Because of the reasons that I explained in the chat. Rename during first `box.cfg{}` is highly untrivial. If you want to go back to that discussion, then sure. I need a decision from products on which way should be selected. The message is in VK Teams in the Scaling Team, 23 March.

Otherwise users will still need at least 2 `box.cfg{}` calls which means that empty->non-empty allowance wouldn't make the setup simpler anyway. But I can allow that too. Again, since this is a behaviour change, I need an approval from a product. (I wouldn't ask some weeks ago, but at this point I have to confirm every public API change.)

@Mons or @sergos or any lead/PM/manager, I need a decision again.

Reply:
> You have to realize that we still can't allow that at the first box.cfg{}. Because of the reasons that I explained in the chat. Rename during first box.cfg{} is highly untrivial.

As for me, I didn't ask about allowing rename in the first box cfg. And I don't remember discussing it in the chat. I only asked about not requiring force_recovery when setting "empty" -> "non-empty". I think it will make the procedure easier than requiring force_recovery on every node.
So the procedure should be:
1. On a **running** cluster choose a master and issue `box.cfg{replicaset_name='smth', cluster_name='smth'}`
2. Wait for the names to reach everyone via replication.
3. On all the replicas set `box.cfg{replicaset_name='smth', cluster_name='smth'}` any time later.
4. If the node is not running, and doesn't know cluster/replicaset name, it has to be started **without** a name first. Later on all the restrictions go avay once it syncs with master and receives a name.

> Otherwise users will still need at least 2 box.cfg{} calls which means that empty->non-empty allowance wouldn't make the setup simpler anyway. But I can allow that too.

I think it would make the procedure easier, because you don't have to set `force_recovery` on everyone and then reset it back to false.

> Again, since this is a behaviour change, I need an approval from a product. (I wouldn't ask some weeks ago, but at this point I have to confirm every public API change.)
>
> @Mons or @sergos or any lead/PM/manager, I need a decision again.

Let's ask @sergos then.

To be crystal clear, I only ask for these changes:
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index a41227394..225aaed89 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -4045,7 +4045,8 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
                                          BOX_SCHEMA_FIELD_VALUE,
                                          field_name) != 0)
                        return -1;
-               if (box_is_configured() && strcmp(name, CLUSTER_NAME) != 0) {
+               if (box_is_configured() && *CLUSTER_NAME != 0 &&
+                   strcmp(name, CLUSTER_NAME) != 0) {
                        if (!box_is_force_recovery) {
                                diag_set(ClientError, ER_UNSUPPORTED,
                                         "Tarantool", "cluster name change "
@@ -4069,7 +4070,8 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
                                          BOX_SCHEMA_FIELD_VALUE,
                                          field_name) != 0)
                        return -1;
-               if (box_is_configured() && strcmp(name, REPLICASET_NAME) != 0) {
+               if (box_is_configured() && *CLUSTER_NAME != 0 &&
+                   strcmp(name, REPLICASET_NAME) != 0) {
                        if (!box_is_force_recovery) {
                                diag_set(ClientError, ER_UNSUPPORTED,
                                         "Tarantool", "replicaset name change "
```

Reply:
Done, PTAL. I also had to allow to subscribe on a master with non-empty replicaset name when the replica has this name locally empty. Otherwise it would require force-recovery in case a replica would reconnect during the name's first time setting and would see the name mismatch (empty locally vs non-empty on master). A test is added as well.

Reply:
Thanks! Nice.